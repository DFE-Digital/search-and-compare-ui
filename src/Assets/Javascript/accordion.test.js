import Accordion from "./accordion"

describe("Accordion module", () => {
  it("exposes a function", () => {
    expect(typeof Accordion).toBe("function")
  })

  let accordion

  beforeEach(() => {
    document.body.innerHTML = `
<div data-module="accordion">
  <div class="govuk-accordion__section">
    <div class="govuk-accordion__section-header">Header 1</div>
    <div class="govuk-accordion__section-body">Body 1</div>
  </div>
  <div class="govuk-accordion__section">
    <div class="govuk-accordion__section-header">Header 2</div>
    <div class="govuk-accordion__section-body">Body 2</div>
  </div>
</div>
`
    accordion = new Accordion(document.querySelector('[data-module="accordion"]'))
  })

  describe("init()", () => {
    describe("when all sections are closed", () => {
      beforeEach(() => {
        accordion.init()
      })

      it("initialises accordion HTML correctly", () => {
        expect(document.body.innerHTML).toMatchSnapshot()
      })
    })

    describe("when one section is open", () => {
      beforeEach(() => {
        document.querySelector(".govuk-accordion__section").classList.add("govuk-accordion__section--expanded")
        accordion.init()
      })

      it("initialises accordion HTML correctly", () => {
        expect(document.body.innerHTML).toMatchSnapshot()
      })
    })

    // describe("when all sections are open", () => {
    //   beforeEach(() => {
    //     document
    //       .querySelectorAll(".govuk-accordion__section")
    //       .forEach($section => $section.classList.add("govuk-accordion__section--expanded"))
    //     accordion.init()
    //   })

    //   it("initialises accordion HTML correctly", () => {
    //     expect(document.querySelector(".govuk-accordion__expand-all").innerHTML).toBe("Close all")
    //     expect(document.body.innerHTML).toMatchSnapshot()
    //   })
    // })
  })

  describe("events", () => {
    describe("clicking on open all", () => {
      beforeEach(() => {
        accordion.init()
        document.querySelector(".govuk-accordion__expand-all").click()
      })

      it("opens all accordion sections", () => {
        document.querySelectorAll(".govuk-accordion__section").forEach($section => {
          expect($section.getAttribute("aria-expanded")).toBe("true")
        })
      })
    })

    describe("clicking on an accordion section header", () => {
      let $sections

      beforeEach(() => {
        accordion.init()
        $sections = document.querySelectorAll(".govuk-accordion__section")
        $sections[0].querySelector(".govuk-accordion__section-header").click()
      })

      it("opens the first accordion section", () => {
        expect($sections[0].getAttribute("aria-expanded")).toBe("true")
      })

      it("does NOT open the second accordion section", () => {
        expect($sections[1].getAttribute("aria-expanded")).toBe("false")
      })
    })

    describe("pressing enter on an accordion section header", () => {
      let $sections

      beforeEach(() => {
        accordion.init()
        $sections = document.querySelectorAll(".govuk-accordion__section")
        $sections[0]
          .querySelector(".govuk-accordion__section-header")
          .dispatchEvent(new KeyboardEvent("keypress", { keyCode: 13 }))
      })

      it("opens the first accordion section", () => {
        expect($sections[0].getAttribute("aria-expanded")).toBe("true")
      })

      it("does NOT open the second accordion section", () => {
        expect($sections[1].getAttribute("aria-expanded")).toBe("false")
      })
    })

    describe("pressing a random key on an accordion section header", () => {
      let $sections

      beforeEach(() => {
        accordion.init()
        $sections = document.querySelectorAll(".govuk-accordion__section")
        $sections[0]
          .querySelector(".govuk-accordion__section-header")
          .dispatchEvent(new KeyboardEvent("keypress", { keyCode: 65 }))
      })

      it("does NOT open any accordion sections", () => {
        $sections.forEach($section => {
          expect($section.getAttribute("aria-expanded")).toBe("false")
        })
      })
    })

    describe("clicking on all accordion section headers in turn", () => {
      beforeEach(() => {
        accordion.init()
        document.querySelectorAll(".govuk-accordion__section-header").forEach($section => $section.click())
      })

      it("opens all sections", () => {
        document
          .querySelectorAll(".govuk-accordion__section")
          .forEach($section => expect($section.getAttribute("aria-expanded")).toBe("true"))
      })

      it("changes 'Open all' button to 'Close all'", () => {
        expect(document.querySelector(".govuk-accordion__expand-all").innerHTML).toBe("Close all")
      })
    })
  })
})
