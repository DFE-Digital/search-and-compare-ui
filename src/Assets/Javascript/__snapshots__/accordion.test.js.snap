// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion module init() when all sections are closed initialises accordion HTML correctly 1`] = `
"
<div data-module=\\"accordion\\" class=\\"with-js\\"><div class=\\"govuk-accordion__controls\\"><button class=\\"govuk-accordion__expand-all\\" aria-expanded=\\"false\\" type=\\"button\\">Open all</button></div>
  <div class=\\"govuk-accordion__section\\" aria-expanded=\\"false\\">
    <div class=\\"govuk-accordion__section-header\\" tabindex=\\"0\\" role=\\"button\\">Header 1<span class=\\"govuk-accordion--icon\\"></span></div>
    <div class=\\"govuk-accordion__section-body\\">Body 1</div>
  </div>
  <div class=\\"govuk-accordion__section\\" aria-expanded=\\"false\\">
    <div class=\\"govuk-accordion__section-header\\" tabindex=\\"0\\" role=\\"button\\">Header 2<span class=\\"govuk-accordion--icon\\"></span></div>
    <div class=\\"govuk-accordion__section-body\\">Body 2</div>
  </div>
</div>
"
`;

exports[`Accordion module init() when one section is open initialises accordion HTML correctly 1`] = `
"
<div data-module=\\"accordion\\" class=\\"with-js\\"><div class=\\"govuk-accordion__controls\\"><button class=\\"govuk-accordion__expand-all\\" aria-expanded=\\"false\\" type=\\"button\\">Open all</button></div>
  <div class=\\"govuk-accordion__section govuk-accordion__section--expanded\\" aria-expanded=\\"true\\">
    <div class=\\"govuk-accordion__section-header\\" tabindex=\\"0\\" role=\\"button\\">Header 1<span class=\\"govuk-accordion--icon\\"></span></div>
    <div class=\\"govuk-accordion__section-body\\">Body 1</div>
  </div>
  <div class=\\"govuk-accordion__section\\" aria-expanded=\\"false\\">
    <div class=\\"govuk-accordion__section-header\\" tabindex=\\"0\\" role=\\"button\\">Header 2<span class=\\"govuk-accordion--icon\\"></span></div>
    <div class=\\"govuk-accordion__section-body\\">Body 2</div>
  </div>
</div>
"
`;
